Add labels and links

itamargiv opened the pull request:
_**Note:** This change depends on #113_

This change adds an entity ID to Wikidata label map to the results page. This map is then used to display labels for various entity IDs displayed on the page.

Bug: [T291569](https://phabricator.wikimedia.org/T291569)

Silvan-WMDE commented on routes/web.php:
```diff
@@ -30,22 +32,37 @@
+            ->whereIn('item_id', $itemIds)
+            ->lazy();
+
+        $entityIds = $mismatches->reduce(function ($ids, $mismatch) {
+            $wikidataValue = $mismatch->wikidata_value;
+            $entityValue = preg_match(config('mismatches.validation.item_id.format'), $wikidataValue);
+
+            if (!in_array($mismatch->property_id, $ids)) {
+                $ids[] = $mismatch->property_id;
+            }
+
+            if ($entityValue && !in_array($wikidataValue, $ids)) {
+                $ids[] = $wikidataValue;
+            }
+
+            return $ids;
+        }, $itemIds);
```
Ok, just to summarise:
this `reduce` function iterates through all the mismatches and collects entityIds for which labels need to be fetched. It also starts with the `$itemIds` as initial value, which means that labels will be fetched for all Q-ids from the input field, regardless of whether mismatches for them can be found in the db or not. Is that right?

I think at least this last point deserves a comment in the code, if not a quick overall explanation as to what the combination of `$entities`, `$mismatches`, the `reduce` function and the `$itemIds` argument will actually achieve.

Also, this is getting quite complex and I'm wondering at which point it actually makes sense to move it all into its own `ResultsController`, rather than adding more and more functionality to a single ->get()` method in the web.php file. Collecting entity IDs and sending them to Wikidata to fetch labels may actually deserve its own method (or two), doesn't it?

itamargiv replied:
> [...], which means that labels will be fetched for all Q-ids from the input field, regardless of whether mismatches for them can be found in the db or not. Is that right?

Correct!

> I think at least this last point deserves a comment in the code, if not a quick overall explanation as to what the combination of $entities, $mismatches, the reduce function and the $itemIds argument will actually achieve.

Good point, will do.

> Collecting entity IDs and sending them to Wikidata to fetch labels may actually deserve its own method (or two), doesn't it?

Yes, agreed.

itamargiv replied:
Done

Silvan-WMDE commented on tests/Feature/WebRouteTest.php:
```diff
@@ -103,6 +103,8 @@ public function test_results_route_retrieves_mismatches()
                 // Casting values to string, as it seems that the inertia
                 // testing helper also converts all values to strings
+                'item_id' => (string) $mismatch->item_id,
+                'statement_guid' => (string) $mismatch->statement_guid,
```
these two are not actually new, they've always been there, right? It's just that now they are needed, and therefore you newly include them in the test?

guergana replied:
I also noticed this

itamargiv replied:
Yes indeed, as long as I don't need them, there is no need to test for them, otherwise tests will become brittle.

Silvan-WMDE commented on routes/web.php:
```diff
@@ -30,22 +32,37 @@
-            'results' => $results,
+            'item_ids' => $itemIds,
+            'results' => $mismatches->groupBy('item_id'),
+            'labels' => $wikidata->getLabels($entityIds, App::getLocale())
```
this whole new `labels` array is added here but never tested for in ẀebRouteTest.php` below, I think.

itamargiv replied:
Good point, I will update the tests.

itamargiv replied:
Done

Silvan-WMDE commented on resources/js/Pages/Results.vue:
```diff
@@ -19,26 +21,47 @@
+            labels: Object as PropType<LabelMap>
+        },
+        methods: {
+            mapLabels(mismatches: Mismatch[]): LabelledMismatch[]{
+                return mismatches.map(mismatch => ({
+                    property_label: this.labels[mismatch.property_id],
+                    value_label: this.labels[mismatch.wikidata_value] || null,
+                    ...mismatch
+                }));
```
ok, this looks equally scary as the `reduce` function below, but basically adds a `property_label` and `value_label` field (or `null`, if not applicable?) to each mismatch, while leaving the rest of the mismatch fields untouched, right? Nice. Maybe also worth a comment? Or do you think the code speaks for itself here?

Also, the method name: wouldn't `addLabels` actually be a better fit here, because `map` is the really technical activity of iterating through the mismatches and applying some function to each element, rather than the conventional "map A to B"?

itamargiv replied:
> Maybe also worth a comment? Or do you think the code speaks for itself here?

To me the code speaks for itself, this is a pretty common pattern in JS. But, I'm not a big fan of implicit knowledge, so if you feel like a comment is needed, that is reason enough for me to add one :smiley: 

> Also, the method name: wouldn't addLabels actually be a better fit here

Sure, yeah, why not all for clarity.

itamargiv replied:
Done

Silvan-WMDE reviewed (changes requested):
Hey, this is fanatastic work. It adds so much functionality into short and precise code, I really like it. just added some comments below, almost only for my own understanding.

Still, I think I found some tests missing, and I guess if we had browser tests for the results page, then they would have to be modified a little with this change? Shouldn't we add some?

itamargiv commented:
Hey hey, thank you for the thorough review I updated the code according to your remarks, added more comments so that hopefully the code is clearer now. I also updated the browser tests, and found a bug with Qids that start with a lowercase Q (see last commit). Feel free to unresolve any of the comments if there is more to add or change.

Silvan-WMDE commented on routes/web.php:
```diff
@@ -30,24 +33,8 @@
-            'results' => $results,
-        ]);
-    })->name('results');
+    ->get('/results', [ResultsController::class, 'index'])
+    ->name('results');
```
ah, doesn't it look great? :-)

itamargiv replied:
It truely does

Silvan-WMDE commented on tests/Vue/Components/MismatchRow.spec.js:
```diff
@@ -27,11 +26,35 @@ describe('MismatchesRow.vue', () => {
             mismatch.import_meta.created_at
         );
     });
+
+    it('shows wikidata label over value when available', () => {
```
nice addition

itamargiv replied:
Thanks :)

Silvan-WMDE reviewed (approved):
Thank you for all the changes! :100:

itamargiv commented:
Thanks for the typo fix!
